### ISSUE TITLE
Add state persistence listener to PluginRegistry.Registrar

### ISSUE THREAD

**Issue by sir-boformer** [flutter/flutter#22328]
I'm not sure about the iOS side of things.

To solve #17950 and to provide convenient solutions to #6827, plugins must have a chance to persist some state before the Android host activity is destroyed. 

On Android, `onSaveInstanceState` on the activity is called before `onDestroy` and allows you to add values to a bundle. State restoration usually happens in the onCreate method of the activity.

I propose the addition of a new listener interface to the `PluginRegistry`:

```java
StatePersistenceListener {
    void onSaveInstanceState(Bundle outState);
    void onRestoreInstanceState(Bundle savedInstanceState);
}
```

Use cases:
-------------

* A router plugin that stores the current route even when the app is killed in background
* Fix for the image_picker so that it works on all Android devices

**Comment by passsy** [comment]
This interface will not help. Calls to Dart and back are **async** whereas `onSaveInstanceState` requires to fill the `Bundle` and return synchronously.

**Comment by sir-boformer** [comment]
@passsy Yes, calls from Dart are async, but that does not make it unusable. You just have to make sure that the state is stored in a field of the platform plugin delegate so that no dart code must be called when the state is persisted.

* In case of a router plugin, you could send the changed route in URL form to the platform when it changes.
* In case of image_picker (bug #17950), [the state that must be persisted](https://github.com/flutter/plugins/blob/master/packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java#L107) is already stored in the platform delegate

**Comment by darshankawar** [comment]
Don't see the proposal being addressed or added here: https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/plugin/common/PluginRegistry.java
